File: part1-coding/06-toposort-course-schedule/java/06ToposortCourseSchedule.java
import java.util.*;
public class TopoSort {
    public static int[] topoSort(int n, int[][] edges){
        List<List<Integer>> g=new ArrayList<>();
        for(int i=0;i<n;i++) g.add(new ArrayList<>());
        int[] indeg=new int[n];
        for(int[] e: edges){ g.get(e[1]).add(e[0]); indeg[e[0]]++; }
        Queue<Integer> q=new LinkedList<>();
        for(int i=0;i<n;i++) if(indeg[i]==0) q.add(i);
        int[] order=new int[n]; int k=0;
        while(!q.isEmpty()){
            int u=q.poll(); order[k++]=u;
            for(int v: g.get(u)) if(--indeg[v]==0) q.add(v);
        }
        return k==n?order:new int[0];
    }
}

File: part1-coding/06-toposort-course-schedule/python/06_toposort_course_schedule.py
from collections import defaultdict, deque
def topo_sort(n, edges):
    g=defaultdict(list); indeg=[0]*n
    for a,b in edges: g[b].append(a); indeg[a]+=1
    q=deque([i for i in range(n) if indeg[i]==0]); order=[]
    while q:
        u=q.popleft(); order.append(u)
        for v in g[u]:
            indeg[v]-=1
            if indeg[v]==0: q.append(v)
    return order if len(order)==n else None
